lib/image_optim/cache.rb

```diff
@@ -0,0 +1,65 @@
+require 'digest/sha1'
+require 'fspath'
+require 'image_optim/cache_path'
+
+class ImageOptim
+  # Handles image cache
+  class Cache
+    def initialize(image_optim, workers_by_format)
+      return unless image_optim.cache_dir
+      @cache_dir = FSPath.new(image_optim.cache_dir)
+      @cache_worker_digests = image_optim.cache_worker_digests
+      @options_by_format = Hash[workers_by_format.map do |format, workers|
+        [format, workers.map(&:inspect).sort.join(', ')]
+      end]
+      @bins_by_format = Hash[workers_by_format.map do |format, workers|
+        [format, workers.map(&:used_bins).flatten!.map! do |sym|
+          bin = image_optim.resolve_bin!(sym)
+          "#{bin.name}[#{bin.digest}]"
+        end.sort!.uniq.join(', ')]
+      end]
+    end
+
+    def fetch(original)
+      return yield unless @cache_dir
+
+      cached = @cache_dir / digest(original, original.format)
+      return cached if cached.file?
+      return if cached.symlink?
+
+      optimized = yield
+
+      if optimized
+        tmp = @cache_dir / optimized.basename
+        FileUtils.mv(optimized, tmp)
+        cached.dirname.mkpath
+        tmp.rename(cached)
+        cached_ = @cache_dir / digest(cached, original.format)
```

Comment:
Confusing variable name, maybe extract a method for marking unoptimizable images?